### PR TITLE
Add new module to split events

### PR DESCRIPTION
**Pull request by legoguy1000**
## What does this PR do?

Replicates the functionality of the httpjson input's split capability.  Created a new parser

Also added the ability parse events of JSON arrays (ex. `[{<data>},{<data>}...]`) to AWS S3 input.

Goal would be to add this to all the inputs so it's a standard capability. The big driver for this was we were ingesting AWS logs offline so had to use the log input with the AWS modules but had issues because the `expand_key...` setting is specific to the AWS S3 input. Also we're ingesting Azure NSG logs in a similar manner which are also multi level nested.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

**Comment by legoguy1000** [comment]
@P1llus After we talked I thought about how to implement a split functionality as agnostically as possible. I copied the HTTP Json input's code. I'm sure it could be much better/more efficient.

**Comment by mr1716** [comment]
Quick question:
Is this a new split processor for Filebeat?

**Comment by legoguy1000** [comment]
> Quick question: Is this a new split processor for native Filebeat to run that is equivalent to the functionality in the Elasticsearch Split processor?

This is to replicate the capability of the split processor from Logstash that can split an event into multiple events. The Elasticsearch split processor just splits a string into an array within the same event.  This is not a processor, its part of the inputs listed above.

**Comment by marc-gr** [comment]
To have a clear picture of how this would look like, this would be part of common input options, similar to tls, looking similar to:

```yaml
input:
   type: mqtt
   # ...
   split:
      # ...
```

and not a new processor, so it will always happen before processors take place, is this assumption correct? Do you think it would make sense to see how feasible would be to have this as a processor? (I am aware the amount of changes to achieve that would be much bigger) If I am not mistaken, the processor form would be closer to what logstash does, as it can be used in combination with other filters if I am not mistaken.

If there is no interest or use case to have this as a processor I am fine for the simpler approach. Would be interesting to have input from @elastic/elastic-agent-data-plane

**Comment by mr1716** [comment]
> To have a clear picture of how this would look like, this would be part of common input options, similar to tls, looking similar to:
> 
> ```yaml
> input:
>    type: mqtt
>    # ...
>    split:
>       # ...
> ```
> 
> and not a new processor, so it will always happen before processors take place, is this assumption correct? Do you think it would make sense to see how feasible would be to have this as a processor? (I am aware the amount of changes to achieve that would be much bigger) If I am not mistaken, the processor form would be closer to what logstash does, as it can be used in combination with other filters if I am not mistaken.
> 
> If there is no interest or use case to have this as a processor I am fine for the simpler approach. Would be interesting to have input from @elastic/elastic-agent-data-plane

@marc-gr I and others would find a lot of value to have it also be a native Filebeat processor as well as this approach.

**Comment by marc-gr** [comment]
> > To have a clear picture of how this would look like, this would be part of common input options, similar to tls, looking similar to:
> > ```yaml
> > input:
> >    type: mqtt
> >    # ...
> >    split:
> >       # ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and not a new processor, so it will always happen before processors take place, is this assumption correct? Do you think it would make sense to see how feasible would be to have this as a processor? (I am aware the amount of changes to achieve that would be much bigger) If I am not mistaken, the processor form would be closer to what logstash does, as it can be used in combination with other filters if I am not mistaken.
> > If there is no interest or use case to have this as a processor I am fine for the simpler approach. Would be interesting to have input from @elastic/elastic-agent-data-plane
> 
> @marc-gr I can submit another enhancement request to make a split processor request.

I'd like input of a team with more context, but I'd assume having one or the other version of it is preferred instead of both. I'd wait to decide how to proceed before having more feedback on this if that makes sense.

**Comment by legoguy1000** [comment]
@marc-gr you're example is correct. I talked to @P1llus about creating a processor originally and spent a good amount of time trying for my self to see what would need to change to enable a processor similar to logstash and being a novice at GoLang came up dry, as u said it would take quite a bit of refactoring. This same concept for a processor has been requested in Elasticsearch for quite a while too, https://github.com/elastic/elasticsearch/issues/56769. The simplest method was to add the capability to the inputs and modify the raw data before the events are published.

**Comment by kvch** [comment]
How about adding it as a `parser` like `multiline` and friends? Ref: https://github.com/elastic/beats/blob/master/libbeat/reader/parser/parser.go#L41
This seems like a good feature and it would be nice to make it available for all inputs.

**Comment by legoguy1000** [comment]
> How about adding it as a `parser` like `multiline` and friends? Ref: https://github.com/elastic/beats/blob/master/libbeat/reader/parser/parser.go#L41 This seems like a good feature and it would be nice to make it available for all inputs.
> 
> Unfortunately, creating a processor that splits events is close to impossible due to the current architecture. I do not expect anyone to refactor the code so it can be implemented.

It definitely could be. Probably best to be added to the ndjson parser but could be its own. Are the parsers able to be used by all the inputs?

**Comment by legoguy1000** [comment]
Actually I think I found an issue with trying to move it to a parser. Here the is the code from the Kafka input where it calls the parsers, https://github.com/elastic/beats/blob/master/filebeat/input/kafka/input.go#L310. Each parser returns a single message and the input expects a single message from the parsers to create a single event.

**Comment by legoguy1000** [comment]
@marc-gr @kvch any additional thoughts?

**Comment by legoguy1000** [comment]
What drive this was data like https://docs.microsoft.com/en-us/azure/network-watcher/network-watcher-nsg-flow-logging-overview#sample-log-records

**Comment by legoguy1000** [comment]
Any thoughts on this?

**Comment by kvch** [comment]
> Actually I think I found an issue with trying to move it to a parser. Here the is the code from the Kafka input where it calls the parsers, https://github.com/elastic/beats/blob/master/filebeat/input/kafka/input.go#L310. Each parser returns a single message and the input expects a single message from the parsers to create a single event.

I am not sure I follow what you mean. If I understood correctly you can just create an special adapter parser, that would create one event from every split event. Could you please share your work so far where you are turning the splitter into a parser?

**Comment by legoguy1000** [comment]
@kvch I haven't had a chance to really work on the parser but I went through the code of the existing parsers. Idk how I could make one that follows the same standard but outputs more than 1 event when every parser and input that uses them expects each parser to return a single event. The only way to modify that would be to update every parser and input to allow multiple events to be returned by parsers. Am I completely missing the mark?

**Comment by kvch** [comment]
You are correct, a parser can only return one event on `Next`. The adapter I have in mind buffers the events and returns them one by one.

Something like this:

```golang 
type splitterReader struct {
   in io.Reader // or whatever you need in your implementation
   ctx context.Context
   cancel context.CancelFunc
   config Config
   buf    chan reader.Message
}

func (*r splitterReader) Next() (reader.Message, error) {
   r.once.Do(func() {
     go r.reading()
   })

   for r.ctx.Err() == nil {
    case msg := <- r.buf
    return msg, nil
  }
}

func (r *splitterReader) reading() {
   for r.ctx.Err() == nil {
       record, err := r.in.Read()
       events := splitEvents(record)
       for _, evt := range events {
          r.buf <- evt
      }
    }
}

func (r *splitterReader) Close() error {
    r.cancel()
}
```

This is not good code above, just something I put together how I would try to implement the splitter.

**Comment by legoguy1000** [comment]
Ok, I can try something like that.

**Comment by jlind23** [comment]
@legoguy1000 any update here?

**Comment by legoguy1000** [comment]
> @legoguy1000 any update here?

I have started to work on the parser method but haven't had too much time to work it.

**Comment by legoguy1000** [comment]
@kvch I pushed my first implementation of the parser method but am getting null pointer errors.  I could definitely use some help with the implementation. I'm testing with the HTTP Endpoint Input.
Filebeat.yml
```
filebeat.inputs:
- type: http_endpoint
  enabled: true
  listen_address: 0.0.0.0
  listen_port: 5555
  parsers:
    - split:
        keep_parent: true
        target: test
        ignore_empty_value: true
```
Error
```
{"log.level":"info","@timestamp":"2022-04-20T20:20:20.240Z","log.origin":{"file.name":"cfgfile/reload.go","file.line":224},"message":"Loading of config files completed.","service.name":"filebeat","ecs.version":"1.6.0"}
2022/04/20 20:20:22 http: panic serving 10.100.100.2:63835: runtime error: invalid memory address or nil pointer dereference
goroutine 98 [running]:
net/http.(*conn).serve.func1()
        net/http/server.go:1802 +0xb9
panic({0x55946b070aa0, 0x55946d1e3c30})
        runtime/panic.go:1047 +0x266
github.com/elastic/beats/v7/libbeat/reader/split.(*splitterReader).Next(0xc000674400)
        github.com/elastic/beats/v7/libbeat/reader/split/split.go:71 +0x9b
github.com/elastic/beats/v7/x-pack/filebeat/input/http_endpoint.(*httpHandler).apiResponse(0xc000476480, {0x55946b63bc18, 0xc00022a540}, 0xc00021c800)
        github.com/elastic/beats/v7/x-pack/filebeat/input/http_endpoint/handler.go:82 +0x2ce
net/http.HandlerFunc.ServeHTTP(0xc0001700b0, {0x55946b63bc18, 0xc00022a540}, 0x0)
        net/http/server.go:2047 +0x2f
github.com/elastic/beats/v7/x-pack/filebeat/input/http_endpoint.(*apiValidationHandler).ServeHTTP(0xc0002b2a80, {0x55946b63bc18, 0xc00022a540}, 0xc00022a540)
        github.com/elastic/beats/v7/x-pack/filebeat/input/http_endpoint/validate.go:131 +0x9b
net/http.(*ServeMux).ServeHTTP(0x0, {0x55946b63bc18, 0xc00022a540}, 0xc0007d4200)
        net/http/server.go:2425 +0x149
net/http.serverHandler.ServeHTTP({0x55946b62b298}, {0x55946b63bc18, 0xc00022a540}, 0xc0007d4200)
        net/http/server.go:2879 +0x43b
net/http.(*conn).serve(0xc000196d20, {0x55946b65e398, 0xc000673410})
        net/http/server.go:1930 +0xb08
created by net/http.(*Server).Serve
        net/http/server.go:3034 +0x4e8
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x559469d4c283]

goroutine 71 [running]:
github.com/elastic/beats/v7/libbeat/reader/split.(*splitterReader).reading(0xc000674400)
        github.com/elastic/beats/v7/libbeat/reader/split/split.go:79 +0x43
created by github.com/elastic/beats/v7/libbeat/reader/split.(*splitterReader).Next.func1
        github.com/elastic/beats/v7/libbeat/reader/split/split.go:68 +0x5f
```
The HTTP request body
```
{
    "test": [
        {"bob": 1234},
        {"bob": 4567}
    ]
}
```
Also currently if I don't define the `parsers` config, the HTTP server never returns a response.  It just hangs.

**Comment by efd6** [comment]
@legoguy1000 From a quick read it looks like you have a nil `split.splitterReader.ctx`. You can check this by changing the `split.New` to be 
```
func New(r reader.Reader, cfg *Config) *splitterReader {
	return &splitterReader{
		reader: r,
		ctx:    context.TODO(),
		cfg:    cfg,
		logger: logp.NewLogger("parser_split"),
		buf:    make(chan reader.Message),
	}
}
```
The end result though should provide a constructor with a context parameter. Conventionally this would be something like this:
```
func New(r reader.Reader, cfg *Config) *splitterReader {
	return NewWith(context.Background(), r, cfg)
}

func NewWith(ctx context.Context, r reader.Reader, cfg *Config) *splitterReader {
	return &splitterReader{
		reader: r,
		ctx:    ctx,
		cfg:    cfg,
		logger: logp.NewLogger("parser_split"),
		buf:    make(chan reader.Message),
	}
}
```
If only the context form is required the background context form can be elided.

**Comment by legoguy1000** [comment]
@efd6 thanks. I'll try that. This is definitely where my Golang skills start to fizzle out a little.

**Comment by efd6** [comment]
The important thing to know here is that context.Context is an interface, so it must not be nil (clients are not required to check its nilness). Worthwhile reading is here: https://go.dev/blog/context.

**Comment by legoguy1000** [comment]
@efd6 I was able to remove the nil pointer but am still having issues with the actual parser processing.  it still just hangs if i define parsers in teh config.

**Comment by efd6** [comment]
@legoguy1000 Is the hang in unit testing for the package or while running as a complete filebeat? If it's in the complete application, can you construct a reproducer in the unit tests?

**Comment by legoguy1000** [comment]
> @legoguy1000 Is the hang in unit testing for the package or while running as a complete filebeat? If it's in the complete application, can you construct a reproducer in the unit tests?

It's as a complete beat. I'm not sure I know how to add that to the unit tests.

**Comment by legoguy1000** [comment]
@efd6 @kvch Any thoughts from you guys?

**Comment by efd6** [comment]
@legoguy1000 If you resolve the conflicts and bring it up-to-date with tip, I can check that out locally tomorrow and take a look. It would be helpful if you could provide instructions for how to reproduce the problem that you are having as well.

**Comment by legoguy1000** [comment]
@efd6 rebased, hopefully i didn't break anything in my rebasing

File: libbeat/common/split/split.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package split
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
```
Comment (kvch) [inline libbeat/common/split/split.go:11]:
```suggestion
	"github.com/elastic/elastic-agent-libs/logp"
```

File: libbeat/common/split/split.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
```
Comment (kvch) [inline libbeat/common/split/split.go:1]: The license must be Apache, not Elastic.

**Comment by kvch** [comment]
Is it ready for review? You have lots of commented out lines, is that intentional? Please remove those.

**Comment by legoguy1000** [comment]
> Is it ready for review? You have lots of commented out lines, is that intentional? Please remove those.

The module I made (not using the parser piece) works and is still in there.  However my attempts to work this into a parser has not been very successful.  I have rough code from what you provided a while ago but haven't gotten it to be successful yet.  So definitely need some assistance on that piece if the parser is the method that we want to go with.

**Comment by efd6** [comment]
Sorry, I got caught up in other things at the end of last week. I'll take a look at the issues today.

**Comment by efd6** [comment]
@legoguy1000 There's a lot of movement here, so it's very hard to see what is going on.

I would suggest backing up a bit, making sure the common/split package is OK, which it should be given that it's a fork of the code in httpjson, then build on the reader/split package with tests to show is is OK and then, wire in one of the inputs here, since http_endpoint is the one that is causing problems, I'd start with that one.

In the shorter term though, can you provide complete instructions to replicate the hang that you see when testing http_endpoint?

File: libbeat/reader/split/split.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+	splitOnce sync.Once
+)
+
+func New(r reader.Reader, cfg *Config) *splitterReader {
+	return &splitterReader{
+		reader: r,
+		cfg:    cfg,
+		logger: logp.NewLogger("parser_split"),
+		buf:    make(chan reader.Message),
+		ctx:    context.TODO(),
```
Comment (kvch) [inline libbeat/reader/split/split.go:63]:
```suggestion
		ctx:    context.Background(),
```

File: libbeat/reader/split/split.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+		split, err := NewSplit(r.cfg, r.logger)
+		if err != nil {
+			return
+		}
+		// We want to be able to identify which split is the root of the chain.
+		split.IsRoot = true
+		data, _ := json.Marshal(message.Content)
+		eventsCh, err := split.StartSplit(data)
+		if err != nil {
+			return
```
Comment (kvch) [inline libbeat/reader/split/split.go:94]: Please log the error here, too.

File: x-pack/filebeat/input/awss3/config.go [hunk 5]
```diff
@@ -127,6 +128,7 @@ type readerConfig struct {
 	LineTerminator           readfile.LineTerminator `config:"line_terminator"`
 	MaxBytes                 cfgtype.ByteSize        `config:"max_bytes"`
 	Parsers                  parser.Config           `config:",inline"`
+	Split                    *split.SplitConfig      `config:"split"`
```
Comment (kvch) [inline x-pack/filebeat/input/awss3/config.go:131]: You have added the new parser named `split`. There is no need for adding `Split` configuration, too.

It should work like this:
```yaml
parsers:
  - split: ~
```

**Comment by kvch** [comment]
> However my attempts to work this into a parser has not been very successful. I have rough code from what you provided a while ago but haven't gotten it to be successful yet. So definitely need some assistance on that piece if the parser is the method that we want to go with.

What is exactly the issue? Let me know if I can help you.

File: x-pack/filebeat/input/awss3/config.go (hunk 5 above)
Comment (legoguy1000) [reply x-pack/filebeat/input/awss3/config.go:131]: That was before I started teh parser work, just haven't removed it yet.  focusing on one input to test with, the http endpoint input. Once its all working, I'll go through and remove the previous implementation of it being an independent config option.

**Comment by legoguy1000** [comment]
I got it kinda working.  IO tested with AWS S3 and Filestream input which already have logic for the parsers.  I've got it splitting a single message but it never moves on to the next line AND when i try to manually stop filebeat (ctl+c), it just hangs.  The setup I have
```
- type: filestream
  id: my-filestream-id
  enabled: true
  paths:
    -<path to logs>/test*.log
  parsers:
    - split:
        keep_parent: true
        target: records
        ignore_empty_value: true
```
log file
```
{"records":[{"systemId":"1"},{"systemId":"2"}]} //This line is split
{"records":[{"systemId":"3"},{"systemId":"4"}]} //It never hits this entry and just hangs
```

**Comment by efd6** [comment]
That looks like a manageable sized reproducer to put into a pure Go test. If you set up a splitter and stream those two events to it, it should reproduce the issue and get you closer to the machine while it's happening. It sounds though like there is lock state being left behind after the first event is processed that's stopping the second one. So where I'd be looking is in making sure you are cleaning up mutexes and channels properly after a split operation has completed.

**Comment by legoguy1000** [comment]
I'll see what I can do.

**Comment by legoguy1000** [comment]
@efd6 It looks like i got it working with both filestream and AWS S3 input..... But now i'm back to filebeat process not stopping when i `CTRL+C`.

**Comment by efd6** [comment]
Not handling a SIGINT sounds like the signal handles logic is being messed up somewhere. The handler for this is in libbeat/service/service.go. Do you see anything in the debug logs when you attempt a cancel? I suspect that you will, but that the cancel func is not being listened to somewhere. The places to look are in event loops and to make sure that things that are waiting on chan inputs have a select case that is receiving from the `ctx.Done()` chan return.

**Comment by legoguy1000** [comment]
> he signal handling logic is being messed up somewhere. The handler for this is in libbeat/service/service.go. Do you see anything in the debug logs when you attempt a cancel? I suspect that you will, but that the cancel func is not being listened to somewhere. The places to look are in event loops and to make sure tha

Ok, i'll take a look.

**Comment by legoguy1000** [comment]
> Not handling a SIGINT sounds like the signal handling logic is being messed up somewhere. The handler for this is in libbeat/service/service.go. Do you see anything in the debug logs when you attempt a cancel? I suspect that you will, but that the cancel func is not being listened to somewhere. The places to look are in event loops and to make sure that things that are waiting on chan inputs have a select case that is receiving from the `ctx.Done()` chan return.

Ok, I'll take a look.  I suspect i'll have to add new logic to pass that into the parser syntax/

**Comment by efd6** [comment]
Yes, that's likely. The convention is to pass the `context.Context` in as `ctx` in the first argument. Then hand it to anyone who wants it and in loops where there is the potential to block, check it, and add a select case for any chan operations.

**Comment by legoguy1000** [comment]
@kvch @efd6 I think i have everything working with the original inputs that already have parsers enabled, AWS S3 and filestream.

**Comment by legoguy1000** [comment]
fixed golang linting

File: filebeat/docs/parsers-options.asciidoc [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+[source,yaml]
+----
+- split:
+	keep_parent: true
+	target: <key to split>
+	ignore_empty_value: true
+	split:
+		keep_parent: true
+		target: <key to split>
```
[the remark this answers is not in the document]
Comment (legoguy1000) [reply filebeat/docs/parsers-options.asciidoc:215]: can do, its currently using the same field names as the httpjson split config, but I agree its a little weird naming currently.

**Comment by kvch** [comment]
Let's split up the parser and the work in the aws input. I think your work in aws is ready to be merged, so we should not block it with the parser work. WDYT?

**Comment by legoguy1000** [comment]
@kvch I removed the AWS piece and I tried to replicate ur issue with the nil pointer and couldn't.  Can u provide the config u used?

**Comment by legoguy1000** [comment]
@kvch can we run this through the CI and see what happens? Also is there a good way to test the changes to the docs?

File: filebeat/input/kafka/input.go [hunk 7]
```diff
@@ -29,8 +29,9 @@ import (
 	"github.com/elastic/beats/v7/libbeat/common/atomic"
 	"github.com/elastic/elastic-agent-libs/mapstr"
 
+	"errors"
```
Comment (efd6) [inline filebeat/input/kafka/input.go:32]: This should be in the std block.

File: filebeat/input/kafka/input.go [hunk 8]
```diff
@@ -254,7 +256,7 @@ func doneChannelContext(ctx input.Context) context.Context {
 }
 
 func (c channelCtx) Deadline() (deadline time.Time, ok bool) {
-	return
+	return //nolint:nakedret //Function is required
```
Comment (efd6) [inline filebeat/input/kafka/input.go:259]:
```suggestion
	return deadline, ok
```

File: filebeat/input/kafka/input.go [hunk 9]
```diff
@@ -308,13 +311,13 @@ func (h *groupHandler) ack(message *sarama.ConsumerMessage) {
 
 func (h *groupHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
 	reader := h.createReader(claim)
-	parser := h.parsers.Create(reader)
-	for h.session.Context().Err() == nil {
+	parser := h.parsers.Create(h.ctx, reader)
+	for errors.Is(h.session.Context().Err(), nil) {
```
Comment (efd6) [inline filebeat/input/kafka/input.go:315]:
```suggestion
	for h.session.Context().Err() == nil {
```

errorlint [should not complain](https://github.com/polyfloyd/go-errorlint/blob/a0c208956a365f0f594df070d05d5359f1492648/errorlint/lint.go#L152-L154) about comparison of errors to nil.

File: filebeat/input/kafka/input.go [hunk 10]
```diff
@@ -308,13 +311,13 @@ func (h *groupHandler) ack(message *sarama.ConsumerMessage) {
...
 		message, err := parser.Next()
-		if err == io.EOF {
+		if errors.Is(err, io.EOF) {
```
Comment (efd6) [inline filebeat/input/kafka/input.go:317]:
```suggestion
		if err == io.EOF { //nolint:errorlint // io.EOF is not wrapped.
```

File: filebeat/input/kafka/input.go [hunk 11]
```diff
@@ -308,13 +311,13 @@ func (h *groupHandler) ack(message *sarama.ConsumerMessage) {
...
 			return nil
 		}
-		if err != nil {
+		if !errors.Is(err, nil) {
```
Comment (efd6) [inline filebeat/input/kafka/input.go:320]:
```suggestion
		if err != nil {
```

File: libbeat/reader/parser/parser.go [hunk 12]
```diff
@@ -86,6 +88,13 @@ func NewConfig(pCfg CommonConfig, parsers []config.Namespace) (*Config, error) {
 	for _, ns := range parsers {
 		name := ns.Name()
 		switch name {
+		case "split":
+			var config split.Config
+			cfg := ns.Config()
+			err := cfg.Unpack(&config)
+			if err != nil {
+				return nil, fmt.Errorf("error while parsing split parser config: %+v", err)
```
Comment (efd6) [inline libbeat/reader/parser/parser.go:96]:
```suggestion
				return nil, fmt.Errorf("error while parsing split parser config: %w", err)
```

File: libbeat/reader/split/split.go [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+	go r.reading()
+	for r.ctx.Err() == nil {
+		msg, ok := <-r.buf
+		if !ok {
+			return reader.Message{}, io.EOF
+		} else {
+			return msg, nil
+		}
+	}
+	return reader.Message{}, nil
```
Comment (efd6) [inline libbeat/reader/split/split.go:71]: Should this not return `r.ctx.Err()`?

File: libbeat/reader/split/split.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+			Content: []byte(maybeMsg.Msg.String()),
+			Bytes:   len([]byte(maybeMsg.Msg.String())),
+		}
+	}
+}
+
+func (r *splitterReader) Close() error {
+	close(r.buf)
+	return r.reader.Close()
+	// return nil
```
Comment (efd6) [inline libbeat/reader/split/split.go:105]: Delete commented out code.

**Comment by legoguy1000** [comment]
ohh interesting situation.  Modeled off of the HTTP JSON Input split functionality, it automatically breaks up JSON arrays into individual events and then the splitting is done for nested data.  THis is an interesting situation on how to handle this.

**Comment by legoguy1000** [comment]
I had a thought today. What if instead of a separate parser, what if I add this to the JSON parser since it requires JSON already?

File: libbeat/reader/split/splitter.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+		child: s,
+	}, nil
+}
+
+func (s *split) StartSplit(raw json.RawMessage) (<-chan maybeMsg, error) {
+	ch := make(chan maybeMsg)
+	var jsonObject mapstr.M
+	if err := json.Unmarshal(raw, &jsonObject); err != nil {
+		return nil, err
+	}
```
Comment (belimawr) [inline libbeat/reader/split/splitter.go:73]: Here you're trying to decode the message as a JSON object (aka a map), but the example @kvch posted here: https://github.com/elastic/beats/pull/29951#issuecomment-1156594940 uses an array, so the decoding fails.

Also if the input file does not contain an empty line at the end, Filebeat just hangs.

**Comment by jlind23** [comment]
Dear contributor, thank you so much for all the energy you put into this PR and we hope that it can be finalized soon.
Until then, as there were no recent updates we will close it in order to keep the repository more manageable for us the maintainers.
Please reopen it as soon as you have time to finish your contribution.

**Comment by legoguy1000** [comment]
no problem, I'll try to address the comments above soon.

**Pull request closed without merging**